File: src/Parrot.Renderers/InputRenderer.cs
namespace Parrot.Renderers
{
    using System;
    using System.Collections.Generic;
    using Parrot.Infrastructure;
    using Parrot.Nodes;
    using Parrot.Renderers.Infrastructure;

    public class InputRenderer : SelfClosingRenderer
    {
        public InputRenderer(IHost host) : base(host)
        {
        }

        public override IEnumerable<string> Elements
        {
            get { yield return "input"; }
        }

        private string GetType(Statement statement, IRendererFactory rendererFactory, IDictionary<string, object> documentHost, object model)
        {
            for (int i = 0; i < statement.Attributes.Count; i++)
            {
                if (statement.Attributes[i].Key.Equals("type", StringComparison.OrdinalIgnoreCase))
                {
                    return RenderAttribute(statement.Attributes[i], rendererFactory, documentHost, model);
                }
            }

            return "hidden";
        }

        public override void Render(IParrotWriter writer, IRendererFactory rendererFactory, Statement statement, IDictionary<string, object> documentHost, object model)
        {
            //get the input type
            string type = GetType(statement, rendererFactory, documentHost, model);
            switch (type)
            {
                case "checkbox":
                case "radio":
                    //special handling for "checked"
                    //is there a checked attribute
                    for (int i = 0; i < statement.Attributes.Count; i++)
                    {
                        if (statement.Attributes[i].Key == "checked")
                        {
                            string attributeValue = RenderAttribute(statement.Attributes[i], rendererFactory, documentHost, model);

                            switch (attributeValue)
                            {
                                case "true":
                                    statement.Attributes[i] = new Nodes.Attribute(statement.Attributes[i].Key, new StringLiteral("\"checked\""));
                                    //.Value = "checked";
                                    break;
                                case "false":
                                case "null":
                                    //remove this attribute
                                    statement.Attributes.RemoveAt(i);
                                    i -= 1;
                                    break;
                            }
                        }
                    }

                    break;
            }

            base.Render(writer, rendererFactory, statement, documentHost, model);
        }
    }
}

File: src/Parrot.Javascript/scripts/Renderers/inputRenderer.ts
///<reference path="../Parser/parser.ts" />
///<reference path="./irenderer.ts" />
///<reference path="./rendererProvider.ts" />
///<reference path="../Infrastructure/ObjectModelValueProvider.ts" />
///<reference path="../Infrastructure/ValueTypeProvider.ts" />
///<reference path="../exceptions.ts" />
///<reference path="./tagBuilder.ts" />
///<reference path="./baseRenderer.ts" />
///<reference path="./selfClosingRenderer.ts" />

class InputRenderer extends SelfClosingRenderer {
    defaultChildTag: string = "";
    rendererProvider: RendererProvider;
    elements: string[] = ["input"];

    render(statement: Statement, host: any[], model: any, rendererProvider: RendererProvider): string {
        //get the local model
        //var localModel = this.getLocalModelValue(host, statement, model);
        var type = this.getType(statement, host, model);
        switch (type) {
            case "checkbox":
            case "radio":
                for (var i = 0; i < statement.attributes.length; i++) {
                    var attribute = statement.attributes[i];
                    if (attribute.key == "checked") {
                        var value = this.renderAttribute(attribute, host, model);
                        switch (value) {
                            case "true":
                                statement.attributes[i] = new Attribute(attribute.key, new StringLiteral("\"checked\"", null, 0));
                                break;
                            case "false":
                            case "null":
                                statement.attributes.splice(i, 1);
                                i -= 1;
                                break;
                        }
                    }
                }
            break;
        }

        return super.render(statement, host, model, rendererProvider);
    }

    getType(statement: Statement, host: any[], model: any) {
        for (var i = 0; i < statement.attributes.length; i++) {
            var attribute = statement.attributes[i];
            if (attribute.key == "type") {
                return this.renderAttribute(attribute, host, model);
            }
        }

        return "hidden";
    }
}